src/Features/Core/Portable/EmbeddedLanguages/RegularExpressions/RegexEmbeddedCompletionProvider.cs

```diff
@@ -0,0 +1,560 @@
+
+        private readonly RegexEmbeddedLanguageFeatures _language;
+
+        public RegexEmbeddedCompletionProvider(RegexEmbeddedLanguageFeatures language)
+        {
+            _language = language;
+        }
+
+        public override bool ShouldTriggerCompletion(SourceText text, int caretPosition, CompletionTrigger trigger, OptionSet options)
+        {
+            if (trigger.Kind == CompletionTriggerKind.Invoke ||
+                trigger.Kind == CompletionTriggerKind.InvokeAndCommitIfUnique)
+            {
+                return true;
+            }
+
+            if (trigger.Kind == CompletionTriggerKind.Insertion)
+            {
+                return IsTriggerCharacter(trigger.Character);
+            }
+
+            return false;
+        }
+
+        private bool IsTriggerCharacter(char ch)
+        {
+            switch (ch)
+            {
+                case '\\': // any escape
+                case '[':  // character class
+                case '(':  // any group
+                case '{':  // \p{
+                    return true;
+            }
+
+            return false;
+        }
+
+        public override async Task ProvideCompletionsAsync(CompletionContext context)
+        {
+            if (!context.Options.GetOption(RegularExpressionsOptions.ProvideRegexCompletions, context.Document.Project.Language))
+            {
+                return;
+            }
+
+            if (context.Trigger.Kind != CompletionTriggerKind.Invoke &&
+                context.Trigger.Kind != CompletionTriggerKind.InvokeAndCommitIfUnique &&
+                context.Trigger.Kind != CompletionTriggerKind.Insertion)
+            {
+                return;
+            }
+
+            var position = context.Position;
+            var (tree, stringToken) = await _language.TryGetTreeAndTokenAtPositionAsync(
+                context.Document, position, context.CancellationToken).ConfigureAwait(false);
+
+            var embeddedContext = new EmbeddedCompletionContext(this, context, tree, stringToken);
+            ProvideCompletions(embeddedContext);
+
+            if (embeddedContext.Items.Count == 0)
+            {
+                return;
+            }
+
+            foreach (var embeddedItem in embeddedContext.Items)
+            {
+                var change = embeddedItem.Change;
+                var textChange = change.TextChange;
+
+                var properties = ImmutableDictionary.CreateBuilder<string, string>();
+                properties.Add(StartKey, textChange.Span.Start.ToString());
+                properties.Add(LengthKey, textChange.Span.Length.ToString());
+                properties.Add(NewTextKey, textChange.NewText);
+                properties.Add(DescriptionKey, embeddedItem.FullDescription);
+                properties.Add(AbstractEmbeddedLanguageCompletionProvider.EmbeddedProviderName, this.Name);
```

Comment:
it is... But it's how the completion API currently works since it's a data-oriented API where youstill want to go from items back to the providers that are responsible for them.  Do youhave an alternate way of doing things here?